File: packages/policy/src/waiver.ts
import type { AssurancePolicy, ControlDefinition, SubjectReference, Waiver } from '@l9/assurance-contracts';
export interface WaiverDecision { readonly valid:boolean; readonly reason:'active'|'expired'|'not-yet-active'|'subject-mismatch'|'control-disallows'|'role-unauthorized'|'duration-exceeded'|'signature-required'; }
export function evaluateWaiver(waiver:Waiver,control:ControlDefinition,subject:SubjectReference,policy:AssurancePolicy,evaluationTime:string):WaiverDecision{
  if(!control.waiver?.allowed)return{valid:false,reason:'control-disallows'};
  if(waiver.controlId!==control.id||!scopeMatches(waiver,subject))return{valid:false,reason:'subject-mismatch'};
  const now=Date.parse(evaluationTime),issued=Date.parse(waiver.issuedAt),expires=Date.parse(waiver.expiresAt);
  if(!Number.isFinite(now)||!Number.isFinite(issued)||!Number.isFinite(expires))return{valid:false,reason:'expired'};
  if(now<issued)return{valid:false,reason:'not-yet-active'}; if(now>=expires)return{valid:false,reason:'expired'};
  if(control.waiver.maximumDurationSeconds!==undefined&&(expires-issued)/1000>control.waiver.maximumDurationSeconds)return{valid:false,reason:'duration-exceeded'};
  const accepted=new Set(policy.waiverAuthorization.acceptedRoles); const required=new Set(control.waiver.requiredRoles??[]); const roles=new Set(waiver.authorizedBy.roles);
  const policyAuthorized=[...roles].some((role)=>accepted.has(role)); const controlAuthorized=required.size===0||[...roles].some((role)=>required.has(role));
  if(!policyAuthorized||!controlAuthorized)return{valid:false,reason:'role-unauthorized'};
  if(policy.waiverAuthorization.requireSignature&&!waiver.signature)return{valid:false,reason:'signature-required'};
  return{valid:true,reason:'active'};
}
export function findActiveWaiver(waivers:readonly Waiver[],control:ControlDefinition,subject:SubjectReference,policy:AssurancePolicy,evaluationTime:string):Waiver|undefined{return[...waivers].sort((a,b)=>a.waiverId.localeCompare(b.waiverId)).find((waiver)=>evaluateWaiver(waiver,control,subject,policy,evaluationTime).valid);}
function scopeMatches(waiver:Waiver,subject:SubjectReference):boolean{const scope=waiver.subjectScope;return scope.repository.host.toLowerCase()===subject.repository.host.toLowerCase()&&scope.repository.owner===subject.repository.owner&&scope.repository.name.replace(/\.git$/i,'')===subject.repository.name.replace(/\.git$/i,'')&&(scope.commit==='*'||scope.commit.toLowerCase()===subject.revision.commit.toLowerCase());}

File: src/l9_assurance/policy/waiver.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from l9_assurance.contracts.time import parse_rfc3339_instant


def evaluate_waiver(
    waiver: Mapping[str, Any],
    control: Mapping[str, Any],
    subject: Mapping[str, Any],
    policy: Mapping[str, Any],
    evaluation_time: str,
) -> dict[str, Any]:
    if not control.get("waiver", {}).get("allowed"):
        return {"valid": False, "reason": "control-disallows"}
    if waiver.get("controlId") != control.get("id") or not _scope_matches(waiver, subject):
        return {"valid": False, "reason": "subject-mismatch"}
    now = parse_rfc3339_instant(evaluation_time)
    issued = parse_rfc3339_instant(waiver.get("issuedAt"))
    expires = parse_rfc3339_instant(waiver.get("expiresAt"))
    if now is None or issued is None or expires is None:
        return {"valid": False, "reason": "expired"}
    if now < issued:
        return {"valid": False, "reason": "not-yet-active"}
    if now >= expires:
        return {"valid": False, "reason": "expired"}
    maximum = control.get("waiver", {}).get("maximumDurationSeconds")
    if maximum is not None and (expires - issued).total_seconds() > maximum:
        return {"valid": False, "reason": "duration-exceeded"}
    accepted_roles = set(policy["waiverAuthorization"]["acceptedRoles"])
    required_roles = set(control.get("waiver", {}).get("requiredRoles", []))
    actual_roles = set(waiver["authorizedBy"]["roles"])
    if not actual_roles.intersection(accepted_roles) or (
        required_roles and not actual_roles.intersection(required_roles)
    ):
        return {"valid": False, "reason": "role-unauthorized"}
    if policy["waiverAuthorization"]["requireSignature"] and not waiver.get("signature"):
        return {"valid": False, "reason": "signature-required"}
    return {"valid": True, "reason": "active"}


def find_active_waiver(
    waivers: Sequence[Mapping[str, Any]],
    control: Mapping[str, Any],
    subject: Mapping[str, Any],
    policy: Mapping[str, Any],
    evaluation_time: str,
) -> dict[str, Any] | None:
    for waiver in sorted(waivers, key=lambda item: item["waiverId"]):
        if evaluate_waiver(waiver, control, subject, policy, evaluation_time)["valid"]:
            return dict(waiver)
    return None


def _scope_matches(waiver: Mapping[str, Any], subject: Mapping[str, Any]) -> bool:
    scope = waiver["subjectScope"]
    return bool(
        scope["repository"]["host"].lower() == subject["repository"]["host"].lower()
        and scope["repository"]["owner"] == subject["repository"]["owner"]
        and scope["repository"]["name"].removesuffix(".git")
        == subject["repository"]["name"].removesuffix(".git")
        and (
            scope["commit"] == "*"
            or scope["commit"].lower() == subject["revision"]["commit"].lower()
        )
    )
